server/src/main/java/org/elasticsearch/index/mapper/RuntimeField.java

```diff
@@ -54,7 +54,7 @@ default XContentBuilder toXContent(XContentBuilder builder, Params params) throw
      * Exposes the {@link MappedFieldType} backing this runtime field, used to execute queries, run aggs etc.
      * @return the {@link MappedFieldType} backing this runtime field
      */
-    MappedFieldType asMappedFieldType();
+    MappedFieldType asMappedFieldType(Function<String, MappedFieldType> lookup);
```

Comment:
For my knowledge, will only runtime aliases override this method? Or do we have other plans to use it?

Also small comment, we could add javadoc for `lookup`.

Reply:
I am not super happy about this method. I added it as I thought we needed it for the object runtime field, but it turns out we won't use it and we have so far all implementors returning `this` which is not useful. This change of signature makes it a bit harder to follow, as it tries to do something specific to aliases (looking up the field type) in a generic way, by introducing a function argument, and this is to prevent a field type from depending on FieldTypeLookup directly, I assume.

Question: do we need the lookup to be dynamic or could it be done at creation time? We still rebuild all mappings at every mapping update don't we?

Reply:
> I am not super happy about this method. I added it as I thought we needed it for the object runtime field, but it turns out we won't use it...

After thinking about this more, having some method like this makes sense to me in general. It seems like `RuntimeField` is a counterpart to `FieldMapper` in that it contains a mapping definition and is able to produce a `MappedFieldType`. I wonder if we could rename the method to `fieldType` to better fit with `FieldMapper`?

As for the function argument, so far I don't see a cleaner approach. To me it's not too strange that runtime fields have access to the concrete field types here. @romseygeek pointed out that we've brainstormed new types of runtime fields that refer to concrete fields -- such as one that combines the contents of multiple text fields like a virtual `copy_to`.